src/d3d9/d3d9_spec_constants.h

```diff
@@ -1,31 +1,119 @@
+
+class D3D9DeviceEx;
+
 namespace dxvk {
 
   enum D3D9SpecConstantId : uint32_t {
-    AlphaCompareOp  = 0,    // Range: 0 -> 7          | Bits: 3
-    SamplerType     = 1,    // 2 bits for 16 samplers | Bits: 32
-                            // ^ not used for vertex shaders
-    FogEnabled      = 2,    // Range: 0 -> 1          | Bits: 1
-    VertexFogMode   = 3,    // Range: 0 -> 3          | Bits: 2
-    PixelFogMode    = 4,    // Range: 0 -> 3          | Bits: 2
-
-    PointMode       = 5,    // Range: 0 -> 3          | Bits: 3
-    ProjectionType  = 6,    // 1 bit for 6 samplers   | Bits: 6
-                            // ^ not supported for vertex shaders
-                            //   PS 1.x only supports up to 6 samplers
-
-    VertexShaderBools = 7,  // 16 bools               | Bits: 16
-    PixelShaderBools  = 8,  // 16 bools               | Bits: 16
-    Fetch4            = 9,  // 1 bit for 16 samplers  | Bits: 16
-                            // ^ not supported for vertex shaders
-
-    SamplerDepthMode  = 10, // 1 bit for 20 samplers  | Bits: 20
-                            // ^ vs + ps
-    SamplerNull       = 11, // 1 bit for 20 samplers  | Bits: 20
-                            // ^ vs + ps
+    SpecSamplerType,        // 2 bits for 16 PS samplers      | Bits: 32
+
+    SpecSamplerDepthMode,   // 1 bit for 20 VS + PS samplers  | Bits: 20
+    SpecAlphaCompareOp,     // Range: 0 -> 7                  | Bits: 3
+    SpecPointMode,          // Range: 0 -> 3                  | Bits: 2
+    SpecVertexFogMode,      // Range: 0 -> 3                  | Bits: 2
+    SpecPixelFogMode,       // Range: 0 -> 3                  | Bits: 2
+    SpecFogEnabled,         // Range: 0 -> 1                  | Bits: 1
+
+    SpecSamplerNull,        // 1 bit for 20 samplers          | Bits: 20
+    SpecProjectionType,     // 1 bit for 6 PS 1.x samplers    | Bits: 6
+
+    SpecVertexShaderBools,  // 16 bools                       | Bits: 16
+    SpecPixelShaderBools,   // 16 bools                       | Bits: 16
+
+    SpecFetch4,             // 1 bit for 16 PS samplers       | Bits: 16
+
+    SpecConstantCount,
+  };
+
+  struct BitfieldPosition {
+    constexpr uint32_t mask() const {
+      return uint32_t((1ull << sizeInBits) - 1) << bitOffset;
+    }
+
+    uint32_t dwordOffset;
+    uint32_t bitOffset;
+    uint32_t sizeInBits;
+  };
+
+  struct D3D9SpecializationInfo {
+    static constexpr uint32_t MaxSpecDwords = 5;
+
+    static constexpr std::array<BitfieldPosition, SpecConstantCount> Layout{{
+      { 0, 0, 32 },  // SamplerType
+      
+      { 1, 0,  20 }, // SamplerDepthMode
+      { 1, 20, 3 },  // AlphaCompareOp
+      { 1, 23, 2 },  // PointMode
+      { 1, 25, 2 },  // VertexFogMode
+      { 1, 27, 2 },  // PixelFogMode
+      { 1, 29, 1 },  // FogEnabled
+
+      { 2, 0,  20 }, // SamplerNull
+      { 2, 20, 6 },  // ProjectionType
+
+      { 3, 0,  16 }, // VertexShaderBools
+      { 3, 16, 16 }, // VertexShaderBools
```

Comment:
This should probably be PixelShaderBools.

Reply:
Fixed.